### PR TITLE
LibGUI+Spreadsheet: Allow custom context menus for Tableviews

### PR DESCRIPTION
**Pull request by kingwill101**

### CODE REVIEW

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp [hunk 1]
```diff
@@ -228,6 +228,28 @@ SpreadsheetView::SpreadsheetView(Sheet& sheet)
         m_table_view->move_cursor(GUI::AbstractView::CursorMovement::Down, GUI::AbstractView::SelectionUpdate::Set);
     };
 
+    m_table_view->row_header().set_context_menu_policy(GUI::ContextMenuPolicy::CustomContextMenu);
+    m_table_view->row_header().on_custom_menu_requested = [this](GUI::ContextMenuEvent& event) {
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetView.cpp:232]: This doesn't feel very intuitive, on two basis:
- It's pretty odd to select a row when its header is _right-clicked_
- Once I set an option for a row like this, I feel like I expect it to apply to further cells created in that row.

I'm not sure what the right way to do this is, but this doesn't feel right.

File: Userland/Libraries/LibGUI/Widget.h [hunk 2]
```diff
@@ -283,6 +289,11 @@ class Widget : public Core::Object {
     void set_accepts_emoji_input(bool b) { m_accepts_emoji_input = b; }
     bool accepts_emoji_input() const { return m_accepts_emoji_input; }
 
+    void set_context_menu_policy(GUI::ContextMenuPolicy policy) { m_context_menu_policy = policy; };
+    ContextMenuPolicy get_context_menu_policy() const { return m_context_menu_policy; }
```
Comment (alimpfard) [inline Userland/Libraries/LibGUI/Widget.h:293]: Getters do not have a `get_` prefix, this should be called `context_menu_policy()`.

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp (hunk 1 above)
Comment (kingwill101) [reply Userland/Applications/Spreadsheet/SpreadsheetView.cpp:232]: Selecting the row on right click I think is required based on the fact that whatever operation you'll be selecting from the context menu would apply to all cells in the row.

I agree though about needing to select additional cells when the row is selected

Comment (kingwill101) [reply Userland/Applications/Spreadsheet/SpreadsheetView.cpp:249]: @alimpfard went ahead and made the requested changes

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp [hunk 3]
```diff
@@ -176,6 +177,13 @@ SpreadsheetView::SpreadsheetView(Sheet& sheet)
             m_table_view->set_column_painting_delegate(last_column_index, make<TableCellPainter>(*m_table_view));
         }
         update_with_model();
+
+        if (m_selected_row.has_value()) {
+            auto result = select_row();
+            if (result.is_error()) {
+                dbgln(result.error().string_literal());
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetView.cpp:184]: Please don't pass non-literals as the format argument of dbgln()

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp [hunk 4]
```diff
@@ -321,6 +354,25 @@ void SpreadsheetView::move_cursor(GUI::AbstractView::CursorMovement direction)
...
 
+ErrorOr<void> SpreadsheetView::select_row()
+{
+    if (!m_selected_row.has_value() || !m_selected_row->is_valid()) {
+        return Error::from_string_literal("Row selected is not valid");
+    }
+
+    Vector<GUI::ModelIndex> indices = {};
+
+    for (auto col_idx = 0; col_idx < m_table_view->model()->column_count(); col_idx++) {
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetView.cpp:365]: Please use full names (`column_index`)

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp [hunk 5]
```diff
@@ -205,7 +213,16 @@ SpreadsheetView::SpreadsheetView(Sheet& sheet)
 
     m_table_view->on_selection_change = [&] {
         m_sheet->selected_cells().clear();
-        for (auto& index : m_table_view->selection().indices()) {
+
+        auto indices = m_table_view->selection().indices();
+
+        if (!indices.is_empty() && m_selected_row.has_value()) {
+            if (m_selected_row->row() != indices.at(0).row() || indices.size() == 1) {
+                (void)m_selected_row.release_value();
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetView.cpp:221]: `m_selected_row.clear()` if that's your intent.

File: Userland/Applications/Spreadsheet/SpreadsheetView.cpp [hunk 6]
```diff
@@ -6,6 +6,7 @@
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetView.cpp:6]: The commit message for this commit seems to have been duplicated?

**Pull request closed without merging**